### DISCUSSION TITLE
Trigger Job on branch created

### DISCUSSION THREAD

**Discussion by dmfilipenko** [community/community#26286] [Actions]
<p>Hello, community.</p>
<p>I didn’t find documentation on how to trigger an action when I’m creating a branch.</p>
<p>What I want, on creating branch <strong>releases/**</strong> to trigger publish. Is it possible right now?</p>

**Comment by cofyc** [comment]
<p>When you push newly created branches to Github. You can watch on push events on the branch. <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onpushpull_requestbranchestags</a></p>

**Comment by BrightRan** [comment, marked as answer]
<p>Hi <a class="mention" href="/dmfilipenko">@dmfilipenko</a> ,</p>
<p>You can use the <strong>Create event</strong> to do this, mode details please reference here:<br>
<a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#create-event-create" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#create-event-create</a><br>
<strong>Note:</strong> When you create a new branch, in the branch which the new branch is based on, the YAML file for the <strong>Create event</strong> workflow must have been existing, otherwise the new branch will not trigger the workflow run.</p>

**Comment by shkhaliq** [comment]
<p>I tried</p>
<pre><code class="lang-auto">on:
  create:
    branches: 
      - releases/*
</code></pre>
<p>and</p>
<pre><code class="lang-auto">
on:
  push:
    types: [created]
    branches:
      - 'releases/**'
</code></pre>
<p>But I am still seeing jobs getting triggered by non-release branches.</p>

**Reply by venkatamutyala** [reply]
This worked for me:

```
on:
  workflow_dispatch:
  push:
    branches:
      - 'feat/**'
      - 'main'
```

**Comment by BrightRan** [comment]
<p><a class="mention" href="/shkhaliq">@shkhaliq</a>,</p>
<p>The <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#create" rel="nofollow noopener"><strong>create</strong></a> event does not support branch filter and tag filter. Anytime you create a new branch or tag, the workflow will be triggered as long as the workflow runs on <strong>create</strong> event has been existing on the base ref of the new branch or tag.</p>
<p>The <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#push" rel="nofollow noopener"><strong>push</strong></a> event does not have any <strong>Activity types</strong>. The activity type “<strong>created</strong>” you set on the push event does not exist.</p>
<p>If your projects really need the feature that the <strong>create</strong> event supports branch filter and tag filter, I recommend that you can directly report a feature request <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>
<p>In addition, as the workaround, you can use the expression <strong>github.ref</strong> to get the name of the created branch from the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github</strong> context</a>, then add an <strong>if</strong> conditional on each job to skip all the jobs in the workflow according to the branch name.<br>
For example:</p>
<pre><code class="lang-yaml">on: create
jobs:
  job1:
    if: ${{ contains(github.ref, 'refs/heads/releases/') }}
    . . .
  job2:
    if: ${{ contains(github.ref, 'refs/heads/releases/') }}
    . . .
  jobN:
    if: ${{ contains(github.ref, 'refs/heads/releases/') }}
    . . .
</code></pre>

**Reply by TWiStErRob** [reply]
Tip: use [startsWith](https://docs.github.com/en/actions/learn-github-actions/expressions#startswith) to clearly express what you mean in the condition.

**Comment by swilliams96** [comment]
<p>Thanks for this <a class="mention" href="/brightran">@brightran</a> this is a really helpful workaround. I’ve submitted feedback from your link as there are still a few issues with this method, which I’ll just call out for anyone else who uses this method instead:</p>
<ol>
<li>The workflow still runs but just skips, instead of not running at all. This pollutes the workflow history with lots of meaningless workflow runs and makes it very difficult to find what you’re looking for (especially for the use-case of this example which I’m also using, which is for release build workflows that actually result in a release).</li>
<li>The filter is not as accurate as you can’t specify a complex <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet" rel="noopener nofollow ugc">filter pattern</a>. I was originally attempting to use <code>release/[0-9]+.[0-9]+.[0-9]+</code> as a filter pattern to target release branches with a 3-part numerical version number. However, this workaround means that this will run on all release branches. Similarly I now can’t specify to <em>not</em> run on release branches that end in <code>-rc</code> in future without a very complex <code>if</code> statement here, for example.</li>
</ol>
<p>It would be great for GitHub to add first-party support for this functionality as it would improve UX and help GitHub actions be a lot more flexible for power users, which would help to achieve feature parity with other build agent competitors and make it more attractive to enterprise users with very complex requirements.</p>

**Comment by dannevesdantas** [comment]
<p>We have also sent a feature request for branch filter suport on the <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#create" rel="noopener nofollow ugc">create</a> event at <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" class="inline-onebox" rel="noopener nofollow ugc">Share feedback - GitHub Support</a>. Hopefully it will arrive soon, so that way the action will not be triggered every time any branch is created.</p>

**Comment by GermanVelibekovHouzz** [comment]
<p>Is there any news regarding built-in support of branches filter by the <a href="https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#create" rel="noopener nofollow ugc">create</a> event?</p>

**Comment by juliamunoz17** [comment]
<p>no filter for branch yet?  :frowning:</p>

**Comment by staticnotdynamic** [comment]
<p>It would tremendously help us achieve our use-cases. I am voicing my concern for the <a href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#delete" rel="noopener nofollow ugc">delete</a> event as well.</p>
<p><a class="mention" href="/dannevesdantas">@dannevesdantas</a> Where is the request you created ?</p>

**Comment by dannevesdantas** [comment]
<p><a class="mention" href="/staticnotdynamic">@staticnotdynamic</a> we have sent it through the old <a href="https://support.github.com/contact/feedback" rel="noopener nofollow ugc">customer feedback form</a> page last year. Looks like the old <a href="https://support.github.com/contact/feedback" rel="noopener nofollow ugc">customer feedback form</a> doesn’t exists anymore and now feature requests should be send via the <a href="https://github.com/github-community/community/discussions" class="inline-onebox" rel="noopener nofollow ugc">Discussions · github-community/community · GitHub</a> repo, by opening a discussion. I believe my feature request was lost when the old <a href="https://support.github.com/contact/feedback" rel="noopener nofollow ugc">customer feedback form</a> was discontinued.</p>

**Comment by staticnotdynamic** [comment]
<p>I have searched for past issues (now discussions) in the new tracker but I couldn’t find something similar to our case. Would you please paste your original proposal there as a discussion ?</p>
<p>If you did, please post back the discussion link so that we can all upvote it.</p>

**Reply by LyKos4** [reply]
https://github.com/community/community/discussions/39831 
I found that there is one opened here.

**Comment by pamle117** [comment]
Do we have filter enabled on branch creation ?

**Comment by RupertBarrow** [comment]
Still hoping for filters on branch creation, like this : 
```
on:
  create:
    branches:
      - "release/**"
```

**Comment by jsoref** [comment]
You can use `push` and then check the created field - https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#push

```yaml
on:
  push:
    branches:
      - 'releases/**'

jobs:
   Something:
      if: github.event.created
```

(thanks to @wizlif for taking the draft and getting it working in https://github.com/orgs/community/discussions/26286#discussioncomment-5826578)

**Reply by wizlif** [reply]
I sampled @jsoref ’s method and it gives me the results i needed, though the workflow history is still polluted with skipped jobs.

```yml
on:
  push:
    branches:
      - 'releases/**'

jobs:
    create-tag:
        if: github.event.created
```

https://github.com/wizlif/test_git_actions/blob/main/.github/workflows/release-tag-with-push.yml

**Reply by srl295** [reply]
I think also this doesn't work with `paths:` - the path won't have changed just by creating the branch.

**Comment by DKnodel** [comment]
Created a new discussion since this one is marked as answered but still no branch filter on the create event.
https://github.com/orgs/community/discussions/54860
Spamming action runs that do nothing for each created feature branch is simply not a solution if you only want to do something for a release branch.

**Comment by hunjixin** [comment]
funny, why not support branch and tag filter. not support yet!!!!!

**Comment by mhulabs** [comment]
Why do we need to push to run a job & why cannot create work on branch / tag. This is counter-intuitive to how other ci/cd does it. I had to search for this specific thread to realize it's not supported by default
